File: backend/app/schemas/document.py
from datetime import datetime
from typing import Optional, List, Any
from pydantic import BaseModel, Field

from app.models.document import OCRMode, DocumentStatus, Importance, BlockType


class BlockBase(BaseModel):
    block_type: BlockType = BlockType.TEXT
    bbox: Optional[List[float]] = None
    text: Optional[str] = None
    table_json: Optional[dict] = None


class BlockResponse(BlockBase):
    id: int
    page_id: int
    block_order: int
    confidence: Optional[float] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class BlockUpdate(BaseModel):
    text: Optional[str] = None
    table_json: Optional[dict] = None


class PageBase(BaseModel):
    page_no: int
    width: Optional[int] = None
    height: Optional[int] = None


class PageResponse(PageBase):
    id: int
    document_id: int
    image_path: Optional[str] = None
    ocr_json: Optional[dict] = None
    raw_text: Optional[str] = None
    layout_score: Optional[float] = None
    confidence: Optional[float] = None
    blocks: List[BlockResponse] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DocumentBase(BaseModel):
    title: str
    department: Optional[str] = None
    doc_type: Optional[str] = None
    importance: Importance = Importance.MEDIUM
    ocr_mode: OCRMode = OCRMode.AUTO


class DocumentCreate(DocumentBase):
    pass


class DocumentUpdate(BaseModel):
    title: Optional[str] = None
    department: Optional[str] = None
    doc_type: Optional[str] = None
    importance: Optional[Importance] = None
    ocr_mode: Optional[OCRMode] = None


class DocumentResponse(DocumentBase):
    id: int
    original_filename: str
    file_path: str
    file_size: Optional[int] = None
    mime_type: Optional[str] = None
    page_count: int = 0
    recommended_ocr_mode: Optional[OCRMode] = None
    precision_score: Optional[int] = None
    status: DocumentStatus
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    processed_at: Optional[datetime] = None
    pages: List[PageResponse] = []

    class Config:
        from_attributes = True


class DocumentListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[DocumentResponse]


class OCRModeRecommendation(BaseModel):
    recommended_mode: OCRMode
    precision_score: int
    reasons: List[str]

File: frontend/src/types/document.ts
export type OCRMode = 'fast' | 'accurate' | 'precision' | 'auto';
export type DocumentStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'review';
export type Importance = 'low' | 'medium' | 'high';
export type BlockType = 'text' | 'table' | 'image' | 'header' | 'footer' | 'list';

export interface DocumentBlock {
  id: number;
  page_id: number;
  block_order: number;
  block_type: BlockType;
  bbox: number[] | null;
  text: string | null;
  table_json: Record<string, unknown> | null;
  confidence: number | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentPage {
  id: number;
  document_id: number;
  page_no: number;
  image_path: string | null;
  width: number | null;
  height: number | null;
  ocr_json: Record<string, unknown> | null;
  raw_text: string | null;
  layout_score: number | null;
  confidence: number | null;
  blocks: DocumentBlock[];
  created_at: string;
  updated_at: string;
}

export interface Document {
  id: number;
  title: string;
  original_filename: string;
  file_path: string;
  file_size: number | null;
  file_type: string | null;
  mime_type: string | null;
  page_count: number;
  department: string | null;
  doc_type: string | null;
  importance: Importance;
  ocr_mode: OCRMode;
  recommended_ocr_mode: OCRMode | null;
  precision_score: number | null;
  status: DocumentStatus;
  error_message: string | null;
  processing_time: number | null;
  created_at: string;
  updated_at: string;
  processed_at: string | null;
  pages: DocumentPage[];
}

export interface DocumentListResponse {
  total: number;
  page: number;
  page_size: number;
  items: Document[];
}

export interface OCRModeRecommendation {
  recommended_mode: OCRMode;
  precision_score: number;
  reasons: string[];
}

export interface DocumentUploadParams {
  file: File;
  title?: string;
  department?: string;
  doc_type?: string;
  importance?: Importance;
  ocr_mode?: OCRMode;
}

export interface ProcessingQueueItem {
  id: number;
  title: string;
  original_filename: string;
  status: DocumentStatus;
  ocr_mode: OCRMode;
  page_count: number;
  file_size: number | null;
  created_at: string;
  updated_at: string;
  processed_at: string | null;
  error_message: string | null;
}

export interface ProcessingQueueResponse {
  total: number;
  pending: number;
  processing: number;
  completed: number;
  failed: number;
  items: ProcessingQueueItem[];
}
